domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+		return errors.Annotate(err, "preparing get protocol ID statement")
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare(`
+DELETE FROM port_range
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, unitEndpointUUID{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		unitEndpointMap := map[string]unitEndpointUUID{}
+		for _, endpoint := range endpointsUnderAction.Values() {
+			var err error
+			unitEndpointMap[endpoint], err = st.getUnitEndpointUUID(ctx, tx, unitUUID, endpoint)
+			if errors.Is(err, UnitEndpointNotFound) {
+				unitEndpointMap[endpoint], err = st.insertUnitEndpoint(ctx, tx, unitUUID, endpoint)
+			}
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for endpoint, unitEndpointUUID := range unitEndpointMap {
+			currentOpenedPorts, err := st.getOpenedPorts(ctx, tx, unitEndpointUUID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			currentNetworkPorts := transform.Slice(currentOpenedPorts, decodePortRange)
+			network.SortPortRanges(currentNetworkPorts)
+
+			resultantOpenPorts := reconcilePorts(currentNetworkPorts, openPorts[endpoint], closePorts[endpoint])
+			network.SortPortRanges(resultantOpenPorts)
+
+			// Return early if no changes are to be made. We previously sorted so
+			// port range permutations aren't picked up as changes
+			if reflect.DeepEqual(currentNetworkPorts, resultantOpenPorts) {
+				continue
+			}
+
+			// Clear out all existing open ports for this unit's endpoint and insert the freshly calculated ones.
+			err = tx.Query(ctx, clearPortRange, unitEndpointUUID).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+
+			for _, portRange := range resultantOpenPorts {
+				var protocolID protocolID
+				err = tx.Query(ctx, getProtocolID, protocolName{Name: portRange.Protocol}).Get(&protocolID)
+				if err != nil {
+					return errors.Annotatef(ProtocolNotValid, "protocol %q", portRange.Protocol)
+				}
+
+				unitPortRange := unitPortRange{
+					ProtocolID:       protocolID.ID,
+					FromPort:         portRange.FromPort,
+					ToPort:           portRange.ToPort,
+					UnitEndpointUUID: unitEndpointUUID.UUID,
+				}
+				err = tx.Query(ctx, insertPortRange, unitPortRange).Run()
+				if err != nil {
+					return domain.CoerceError(err)
+				}
+			}
+		}
+		return nil
+	})
+}
+
+// getUnitEndpointUUID returns the UUID of the specified unit's endpoint.
+// If the unit_endpoint does not exist, UnitEndpointNotFound is returned.
+func (st *State) getUnitEndpointUUID(ctx context.Context, tx *sqlair.TX, unitUUID, endpoint string) (unitEndpointUUID, error) {
+	var result unitEndpointUUID
+	m := sqlair.M{
+		"unit_uuid": unitUUID,
+		"endpoint":  endpoint,
+	}
```

Comment:
Make a type `unitUUIDEndpoint`, this makes it easier to understand what you're passing because it's no longer dynamic.

Reply:
Done here as well. There are now no more `sqlair.M`s